updated error handling sytem with unified way to show user error messeges

nitinawari opened the pull request:
fixed #190 
 - implemented unified way to show error to user with toast message 
404 page 
![image](https://github.com/user-attachments/assets/eda09154-047f-490f-95c1-60f4867b0eb0)

nitinawari commented:
@arkid15r Sir, I have created a complete error-handling system to display user error messages in a unified way. Please let me know if I have missed anything, made any mistakes, or if there are any changes or deletions needed

Rajgupta36 commented:
hiii @nitinawari , update your test cases

harsh3dev commented on frontend/src/pages/Chapters.tsx:
```diff
@@ -30,6 +33,9 @@ const ChaptersPage = () => {
       url: chapter.idx_url,
     }
+    if (error) {
+      return <ErrorDisplay error={error} onRetry={error.action === 'retry' ? retry : undefined} />
```
As the tests are failing, adding a basic test for each of the pages for the error display component should work.

nitinawari replied:
yes i writing this test case but one test is failed just resolving that 
![image](https://github.com/user-attachments/assets/51651da7-8806-4813-b69a-2c4a8c2f6639)

Rajgupta36 reviewed (changes requested):
hiii @nitinawari  can you update the code with latest test case changes

nitinawari commented:
"@Rajgupta36, I’m facing issues with some test cases, all of which are related to a single functionality. Should I update the code without passing these test cases now? Could you help me resolve this issue?"

Rajgupta36 commented:
> "@Rajgupta36, I’m facing issues with some test cases, all of which are related to a single functionality. Should I update the code without passing these test cases now? Could you help me resolve this issue?"

yes nitin. you can share the problems with us in community or here

> "@Rajgupta36, I’m facing issues with some test cases, all of which are related to a single functionality. Should I update the code without passing these test cases now? Could you help me resolve this issue?"

Could you share more details about the specific issues you're facing? maybe i could help .

nitinawari commented:
@Rajgupta36  it is about retry functionality  when user reach 3 times retry attempt it shoud show the  home button but it not working i am unable to debug i write proper functionality but what is missing i am unable to find  out  here is failed test cases should share code 
![image](https://github.com/user-attachments/assets/c8e1d853-1149-403c-8449-324e25c6d174)
![image](https://github.com/user-attachments/assets/0f8925a5-649e-40fb-b518-181ba6fe0e53)

Rajgupta36 commented:
> it is about retry functionality when user reach 3 times retry attempt it shoud show the home button but it not working i am unable to debug i write proper functionality but what is missing i am unable to find out here is failed test cases

Can you share the debugging logs?

nitinawari commented:
@Rajgupta36 there is nothing in dubug log just a retry increament count  but i have this test failure
![image](https://github.com/user-attachments/assets/ca391278-651c-4edb-b9b0-20de74ee031f)

yashpandey06 commented:
> @Rajgupta36 there is nothing in dubug log just a retry increament count but i have this test failure ![image](https://private-user-images.githubusercontent.com/99824048/399176827-ca391278-651c-4edb-b9b0-20de74ee031f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0OTQxOTEsIm5iZiI6MTczNTQ5Mzg5MSwicGF0aCI6Ii85OTgyNDA0OC8zOTkxNzY4MjctY2EzOTEyNzgtNjUxYy00ZWRiLWI5YjAtMjBkZTc0ZWUwMzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDE3MzgxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwNDRiOWNlMjFlMGUzNjRiYTc0ZmU1MDIyM2JlZDgzODI4YmU4ZDAzMGY2MmY3YzBlZTcxYmVjZTQyNTlhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4nWkaU0iPv60yP-4P88Oky6OYdMcDi0UZKnNx6wZt38)

@nitinawari ...should i suggest something ?....first ..please lets try to to add some console.log ststement and check how many times useErrorhandler instantiated and how many times the instance of ErrorService (btw naming is wrong inyour case) is getting created and calling handlerror function.

yashpandey06 commented:
as you have given maximum retry count as 3 ...i am quite sure its not 3 .

nitinawari commented:
i think  here race condition happening now looking into that part also what is meant bu this 
" ErrorService (btw naming is wrong inyour case)"

yashpandey06 commented:
> i think here race condition happening now looking into that part also what is meant bu this " ErrorService (btw naming is wrong inyour case)"

@nitinawari i mean the file name is wrong right ..?


exactly ..that what was i thinking ...

You can try this :
 make  your service file singleton or use mutex .

nitinawari commented:
@yashpandey06 can you plz elaborate more. it is not a problem related to naming or instance it more like to retry attempt logic

yashpandey06 commented:
> @yashpandey06 can you plz elaborate more. it is not a problem related to naming or instance it more like to retry attempt logic

@nitinawari I wrote btw ( by the way) which means it's a suggestion which is not related to current discussion  ...the problem might be related to Error service class object creation in usehandler file.

nitinawari commented:
@yashpandey06  ohh , but that is not a problem i also handle that part 
```
 const handleError = useCallback(
    (err: unknown) => {
      setIsRetrying(false)
      const errorConfig = errorService.handleError(err, context)

      // Increment retry count before checking for max retries
      const newRetryCount = retryCount + 1
      setRetryCount(newRetryCount)

      // Check for max retries
      if (newRetryCount > 3) {
        setError({
          code: 'MAX_RETRIES_EXCEEDED',
          title: 'Too Many Attempts',
          message: 'Please try again later or return home.',
          source: 'runtime',
          action: 'home',
        })
        return
      }

      setError(errorConfig)
    },
    [context, retryCount]
  )
```

Rajgupta36 commented:
> @Rajgupta36 there is nothing in dubug log just a retry increament count but i have this test failure ![image](https://private-user-images.githubusercontent.com/99824048/399176827-ca391278-651c-4edb-b9b0-20de74ee031f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1MDA2NTksIm5iZiI6MTczNTUwMDM1OSwicGF0aCI6Ii85OTgyNDA0OC8zOTkxNzY4MjctY2EzOTEyNzgtNjUxYy00ZWRiLWI5YjAtMjBkZTc0ZWUwMzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDE5MjU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwZjc4YWVjZTVhNDM1ZGZjNWI1MGYzNDUwYjhkZWNiZDdhZjQyZDhjYzk5MGY5YWQzMmJmN2FjZjdlZjUwNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zUzByx9gtntsV8SXRRkaSYqDHGoehb6_vnzeJ-3VHRs)

It shows another error (search error) triggered before this one. I think in the above logs, you can see the exact component that is currently rendered and identify the difference between your test case and the received result.

Rajgupta36 commented:
@nitinawari, could you please commit your test cases .maybe I can help you

nitinawari commented:
Hi @Rajgupta36, sorry, but I'm out of the house right now and don't have my laptop. I will commit tomorrow.

arkid15r commented on frontend/src/lib/ErrorServeice.ts:
```diff
@@ -0,0 +1,223 @@
+          action: 'retry',
+        },
+        {
+          code: 'ALGOLIA_CONFIG_ERROR',
+          title: 'Config Error',
+          message: 'Algolia keys not found.',
+          source: 'algolia',
+          action: 'retry',
+        },
```
I'm not sure if we want to show this to user. It should never happen.

nitinawari replied:
In my first PR, I only handled 404, 500, and JavaScript errors. However, later we started fetching data directly from Algolia on the frontend. That’s why I added this part—to allow developers to see errors directly on the screen, not for users. I will remove this part, as well as the unnecessary elements like authentication and rate-limiting issues, which are not aligned with our current Nest.js project

arkid15r commented on frontend/src/lib/ErrorServeice.ts:
This looks quite comprehensive, when I originally created the tasks I planned it to use a much simpler approach -- just some sort of a toast that indicates something went wrong.

I appreciate you implementing it and covering more cases however personally to me it seems a bit too complex (my frontend/React skills not that great). I totally understand 404 page though :)

Maybe it'd be better to start with more simple implementation without sophisticated retry logic and `useSearchPage` refactoring + `error` props drilling (I hope the term is right). 

I'm not sure what to suggest here as I don't want to discourage you and look like I don't appreciate your work. I do and you did a great job. Perhaps, I need other opinions on this. Would you like to add your perspective on why you decided to make it that way?

nitinawari replied:
Yeah, I think I made the task unnecessarily complex by not aligning it with the actual issue 😅. That’s why it took me so much time. I will remove the unnecessary parts and extra functionalities like retry logic, Algolia searching, and auth capture. I initially added them with the intention of making the solution future-proof and production-ready as Nest grows.

However, now I am more clear. Instead of creating multiple extra files like` ErrorService`, `UseErrorHandler`, and `ErrorDisplay`, I’ll keep it simple with a single file focused on the actual errors. This approach will make it clearer and easier for developers to read.

This time, I’ll only address the key issues, such as 404, 500, 503, and some JavaScript runtime errors, and simplify the test cases. I apologize for wasting so much time, and thank you for your understanding.

Rajgupta36 replied:
Hi @nitinawari, I think you should implement 404 and 500 error pages, and for other errors, you can display messages like "Something went wrong" or "Unexpected error." Also, since useSearchPage is used for API calls, you can handle other errors on that page, as Arkid suggested. and also for toast you can take ref from :- https://ui.shadcn.com/docs/components/toast  (note :- not use ui library)

nitinawari replied:
@Rajgupta36 "not use ui library" what do you mean should i avoid using package like  **react-hot-toast**

Rajgupta36 replied:
nah . i mean  shadcn, mui .

arkid15r reviewed (commented):
@nitinawari thanks for addressing this! I looked at the PR and wrote some comments below. I'm mostly interested in other frontend contributor's opinions on the complexity of this approach -- maybe we can simplify it and that would resolve the failing tests issue itself.

@harsh3dev @Rajgupta36 any suggestions from your side? Should we simplify it or keep the architecture as is and focus on fixing the tests?

Thank you!

nitinawari commented:
@arkid15r, thank you for the review! I now understand my assignment in a much simpler way.

nitinawari commented:
@arkid15r Sir, I have removed the unnecessary elements and included only the required functionality in a single file. This file handles 404 and 500 errors, displays a toast message saying 'Something went wrong' for other errors, and also captures errors in Sentry. I hope this implementation meets our requirements. Please let me know if any changes are needed. I am ready to make them

Rajgupta36 commented:
hiii @nitinawari , can you use shadcn toast component instead of react toast ?

nitinawari commented:
Hey @arkid15r,

This PR has unintentionally grown larger than expected. We started integrating ShadCN, and I’ve added ShadCN components because they are useful for error handling in this PR, specifically the toast component. As a result, I also added the necessary ShadCN configurations and other components.

The main focus of this PR should be on `utils/ErrorHandler.tsx`, as it handles the entire process of showing errors to the user in a unified way.

Please review this PR when you have a chance. Thank you!

Rajgupta36 commented on frontend/src/lib/algoliaClient.ts:
```diff
@@ -3,7 +3,8 @@ import { ALGOLIA_APP_ID, ALGOLIA_SEARCH_KEY } from 'utils/credentials'
   if (!ALGOLIA_APP_ID || !ALGOLIA_SEARCH_KEY) {
-    throw new Error('Algolia keys not found.')
+    const error = new Error('something went wrong ')
+    throw error
```
not throw error on screen
code:- 
` if (true) {
    const error = new Error('something went wrong ')
    throw error
  }`
![Screenshot 2025-01-02 234628](https://github.com/user-attachments/assets/aba1e32f-5708-4b13-a61b-d370c6ac28cb)

nitinawari replied:
ok i will resolve

Rajgupta36 reviewed (changes requested):
Also when i throw custom error from this function `fetchAlgoliaData`
it show this behaviour 

https://github.com/user-attachments/assets/d3790bfa-64ad-4c3f-b746-d689c0dfe70b

as well as light mode works well but not dark mode 
![image](https://github.com/user-attachments/assets/dc9c61ce-c6cb-4086-970e-3d3331047635)
![image](https://github.com/user-attachments/assets/eda679f7-91e3-410f-b0c6-e3be8c071e37)

nitinawari commented:
@arkid15r  @Rajgupta36, I’ve resolved all the bugs. I hope everything works fine now. For now, I’ve removed the redirection part because we don’t have a home page yet. If we redirect now, it will create an infinite loop

nitinawari commented:
@Rajgupta36 @arkid15r Could you please review this PR today if you have time? I’ve resolved all the bugs, and everything is working perfectly on my end, capturing the required errors.

arkid15r commented on frontend/tsconfig.app.json:
I removed this file recently. Do we need it?

nitinawari replied:
ok i will remove

arkid15r commented on frontend/tsconfig.json:
```diff
@@ -17,6 +17,7 @@
     "inlineSources": true,
     "baseUrl": ".",
     "paths": {
+      "@/*": ["./src/*"],
```
We already have "*" -- isn't that enough?

nitinawari replied:
this is require for shadcn to add componet in componet/ui folder

arkid15r commented on frontend/vite.config.ts:
```diff
@@ -20,6 +20,7 @@ export default defineConfig({
   ],
   resolve: {
     alias: {
+      '@': path.resolve(__dirname, './src'),
```
Same here.

nitinawari replied:
hey @arkid15r  this changes happen when i added shadcn component . this configuration needed for shadcn

arkid15r reviewed (changes requested):
Please look into these comments when you get a chance:

nitinawari commented on frontend/src/components/Header.tsx:
this file  changes when i run precommit

nitinawari commented on frontend/src/lib/hooks/use-toast.ts:
```diff
@@ -0,0 +1,190 @@
+  UPDATE_TOAST: 'UPDATE_TOAST',
+  DISMISS_TOAST: 'DISMISS_TOAST',
+  REMOVE_TOAST: 'REMOVE_TOAST',
+} as const
```
this is  shadcn component and it's functionality by deafult from library . i didn't make any changes here

nitinawari commented:
Hey @arkid15r, most of the changes occurred here because I added the ShadCN library for toasting

nitinawari commented on frontend/src/lib/hooks/use-toast.ts:
```diff
@@ -0,0 +1,190 @@
+'use client'
```
this is from library

arkid15r commented:
> Hey @arkid15r, most of the changes occurred here because I added the ShadCN library for toasting and most of the configurations, filenames are generated by shadcn

I have two questions here:
  - I'm not sure -- was there any command you ran and it resulted in adding those settings?
  - Would it work w/o those settings?

nitinawari commented:
@arkid15r, 
-yes, to add ShadCN to our project, we need to run some predefined commands, which will set up their configuration to add separate reusable components like Toast.tsx. Their naming conventions are predefined; we just need to import and use them.
https://ui.shadcn.com/docs/installation/vite
 - Additionally, our configuration is also working fine with it

nitinawari commented on frontend/components.json:
This is the ShadCN configuration that stores the predefined settings, such as fonts, default component styling, and aliases. These are needed for ShadCN components

harsh3dev replied:
This one is needed for the Shadcn configurations as it dynamically adds component based on the need

nitinawari commented:
@Rajgupta36 @harsh3dev

harsh3dev commented:
@arkid15r 
The shadcn library adds some config files and It has its own naming convention for example: `use-toast.tsx` and I think we can use the normal import as we are using in the project no need of @.
Also I think we can change the naming of files according to our naming convention @Rajgupta36 can confirm.

Rajgupta36 commented:
![image](https://github.com/user-attachments/assets/8d6f86ca-9793-4a91-8e8b-4f8361cd1e35)

can you make it little bit better.

nitinawari commented on frontend/src/lib/api.ts:
```diff
@@ -1,28 +1,37 @@
+    }
+    return await response.json()
+  } catch (error) {
+    handleAppError(error)
+    throw error
```
we don't need to throw again,Erroewrapper handel everthing  i think i forget to remove that part . it used that part in manual testing

nitinawari commented:
@Rajgupta36  ok i will use new york style and what about messege should i also update message

Rajgupta36 commented:
> @Rajgupta36 ok i will use new york style and what about messege should i also update message
no

nitinawari commented:
@arkid15r Sir, I have fixed the bugs as per your suggestions.
Initially, this PR was solely about error handling. However, we later decided to use ShadCN as our component library. Due to this decision, I needed to add a toast component to display user errors. As a result, this PR includes the ShadCN configuration.

**This PR can be divided into two parts:**

**ShadCN Integration and Configuration**
- The configuration required to set up ShadCN has been added to the project.
- We can easily customize every part of ShadCN, including naming conventions, styling, and other aspects.
 @Rajgupta36  @harsh3dev  can you plz help us in this part to review 

**Error Handling in the Frontend**
- The core part of this PR handles errors in a unified manner.
- Errors are displayed to the user using the toast component provided by ShadCN.
- The focus should be on` lib/ErrorWrapper` and the API calls that have been rewritten for better error handling.

arkid15r reviewed (commented):
@nitinawari I'm going to review this PR, could you resolve the conflicts?

Rajgupta36 commented:
@nitinawari, could you please take a look at the requested changes . i added a shadcn config so you don't need this @ character for importing and exporting

nitinawari commented:
@arkid15r: Ready for review. I hope we can close this today. I have made the requested changes. If you find any bugs or need any improvements, I am happy to make the changes. I am available the whole day.

Rajgupta36 commented:
looks good @nitinawari

arkid15r commented on frontend/src/lib/hooks/useToast.ts:
```diff
@@ -0,0 +1,185 @@
+  REMOVE_TOAST: 'REMOVE_TOAST'
+}
+
+type Action =
+  | {
+      type: ActionType['ADD_TOAST']
+      toast: ToasterToast
+    }
+  | {
+      type: ActionType['UPDATE_TOAST']
+      toast: Partial<ToasterToast>
+    }
+  | {
+      type: ActionType['DISMISS_TOAST']
+      toastId?: ToasterToast['id']
+    }
+  | {
+      type: ActionType['REMOVE_TOAST']
+      toastId?: ToasterToast['id']
+    }
```
Could you explain this syntax to me?

nitinawari replied:
This functionality is useful for real-time updates to toasts but is not needed currently. For now, it has been disabled by removing the ActionType function, and it can be reintroduced later if required

arkid15r reviewed (approved):
This still looks too big to me but now it's easier to understand.
A lot of work, thank you @nitinawari 

I'll bump the estimated points for this task

Let's merge it 👍

nitinawari commented:
@arkid15r  Thank you for appreciating my work and increasing the points for this task! I'll continue to improve and refine it in the future. 😊
